src/libraries/System.Net.Mail/src/System/Net/Mail/SmtpCommands.cs

```diff
@@ -641,6 +641,46 @@ internal static bool Send(SmtpConnection conn, string to, out string response)
         }
     }
 
+    internal static class QuitCommand
+    {
+        private static void CheckResponse(SmtpStatusCode statusCode, string serverResponse)
+        {
+            switch (statusCode)
+            {
+                case SmtpStatusCode.ServiceClosingTransmissionChannel:
+                    {
+                        return;
+                    }
+                case SmtpStatusCode.CommandUnrecognized:
+                default:
+                    {
+                        if ((int)statusCode < 400)
+                        {
+                            throw new SmtpException(SR.net_webstatus_ServerProtocolViolation, serverResponse);
+                        }
+
+                        throw new SmtpException(statusCode, serverResponse, true);
+                    }
+            }
+        }
+
+        private static void PrepareCommand(SmtpConnection conn)
+        {
+            if (conn.IsStreamOpen)
+            {
+                throw new InvalidOperationException(SR.SmtpDataStreamOpen);
+            }
+
+            conn.BufferBuilder.Append(SmtpCommands.Quit);
+        }
+
+        internal static void Send(SmtpConnection conn)
+        {
+            PrepareCommand(conn);
+            SmtpStatusCode statusCode = CheckCommand.Send(conn, out string response);
+            CheckResponse(statusCode, response);
```

Comment:
I'd like to leave reading the response out for now. I'm concerned that this will affect scalability in any apps using `SmtpClient` in async workflows -- blocking here could blow up the thread pool which can be non-obvious for users to diagnose.

I'd support a `QuitAsync` API in a new issue that waits for the response and supports cancellation.

Reply:
Are you suggesting to add an overload

```cs
void CheckCommand.Send(conn)
```

And in future with `QuitAsync` implement async pattern?(like other command i.e. `RecipientCommand` they are using Begin/End pattern at the moment but we'll think how to accomodate)

Reply:
Or simply flush the conn?

Reply:
Added flush.